### PR TITLE
Update CFamily Analyzer to 6.47.0.62356 and add support for MISRA rules

### LINKED ISSUE #4571: Update embedded CFamily analyzer to 6.47.0.62356
6.46.0.62229 → 6.47.0.62356: ~43 new MISRA 2023 rules

[Release notes](https://sonarsource.atlassian.net/issues/?jql=project%20%3D%2010166%20AND%20fixVersion%3D%206.47%20%20ORDER%20BY%20priority%20DESC%2C%20key%20ASC)

Note: additional work will be required to support executing MISRA rules because the MISRA rule metadata is stored separately.
See https://github.com/SonarSource/sonarlint-visualstudio/issues/4604

### PR DESCRIPTION
**Pull request by duncanp-sonar**
NOTE: all of the MISRA rules are disabled by default.

They need to be enabled explicitly to run in Standalone mode, or enabled in a Quality Profile in Connected Mode (such as in the new built-in `Mission critical` Quality Profile for C++)

Fixes #4571 and #4604

### CODE REVIEW

File: build/tools/CFamilyJarPreProcessor/Preprocessor.cs [hunk 1]
```diff
@@ -36,6 +36,7 @@ internal class Preprocessor
         {
                 "Sonar_way_profile.json",
                 "RulesList.json",
+                "MisraRulesList.json",
```
Comment (duncanp-sonar) [inline build/tools/CFamilyJarPreProcessor/Preprocessor.cs:39]: Update the pre-processor so it fetches the additional files we need:
* `MisraRulesList.json`, which lists all of the MISRA rules (not whether they are active or not)
* the rule .json files form the `misra23` subfolder (line #49 below)

File: src/CFamily.UnitTests/Rules/RulesLoaderTest.cs [hunk 2]
```diff
@@ -36,7 +36,8 @@ public class RulesLoaderTest
         public void Read_Rules()
         {
             var rulesLoader = CreateTestSubject();
-            rulesLoader.ReadRulesList().Should().HaveCount(410);
+            // 410 in RulesList.json + 2 in MisraRulesList.json
+            rulesLoader.ReadRulesList().Should().HaveCount(410 + 2);
```
Comment (duncanp-sonar) [inline src/CFamily.UnitTests/Rules/RulesLoaderTest.cs:40]: Added another dummy `MisraRulesList.json` file for this tests, which lists two rules (see next file).

File: src/CFamily.UnitTests/Rules/CFamilySonarWayRulesConfigProviderTests.cs [hunk 3]
```diff
@@ -32,9 +32,13 @@ public class CFamilySonarWayRulesConfigProviderTests
         // Rule data for files in Rules\TestResources\RulesMetadataCache
         private const int Active_C_Rules = 3;
         private const int Inactive_C_Rules = 3;
+        private const int Active_Misra_C_Rules = 1;
+        private const int Inactive_Misra_C_Rules = 1;
 
         private const int Active_CPP_Rules = 4;
         private const int Inactive_CPP_Rules = 2;
+        private const int Active_Misra_Cpp_Rules = 1;
+        private const int Inactive_Misra_Cpp_Rules = 1;
```
Comment (duncanp-sonar) [inline src/CFamily.UnitTests/Rules/CFamilySonarWayRulesConfigProviderTests.cs:41]: Fixing up the existing tests.
I added a dummy json files for four rules (C/C++, Active/Inactive), and listed them in a dummy `MisraRulesList.json` file.

File: src/CFamily/Rules/RulesLoader.cs [hunk 4]
```diff
@@ -40,7 +40,10 @@ public IEnumerable<string> ReadRulesList()
         {
             var rulesList = LoadCFamilyJsonFile<List<string>>("RulesList.json");
             Debug.Assert(rulesList != null, "The CFamily RulesList.json should exist and not be empty");
+            var misraRulesList = LoadCFamilyJsonFile<List<string>>("MisraRulesList.json");
+            Debug.Assert(rulesList != null, "The CFamily Misra RulesList.json should exist and not be empty");
 
+            rulesList.AddRange(misraRulesList);
```
Comment (duncanp-sonar) [inline src/CFamily/Rules/RulesLoader.cs:46]: The CFamily pre-processor copies all of the MISRA files to the `lib` folder.
The only other thing we need to do is combine the lists of MISRA and non-MISRA rules so the rules provider knows the complete set of rules to look at.

File: src/Integration.Vsix.UnitTests/CFamily/CFamilyEmbeddedSonarWayRulesTests.cs [hunk 5]
```diff
@@ -44,12 +44,14 @@ public class CFamilyEmbeddedSonarWayRulesTests
...
+        // Rule data for C-Family plugin v6.47.0.62356
 
-        private const int Active_CPP_Rules = 435;
-        private const int Inactive_CPP_Rules = 165;
+        // Note: does not include MISRA rules - see #4604
+        private const int Active_C_Rules = 210;
+        private const int Inactive_C_Rules = 123;
+
+        private const int Active_CPP_Rules = 436;
+        private const int Inactive_CPP_Rules = 207;
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/CFamily/CFamilyEmbeddedSonarWayRulesTests.cs:54]: image


The CFamily plugin defines 436 + 207 = 643 rules.

The "extra" six rules are the now-deprecated "Common" rules, which are all inactive.

image

That explains the discrepancy in the inactive rule count: between the two screen shots. 207 inactive C++ + 6 inactive Common = 213 inactive in total.

File: src/Integration.Vsix.UnitTests/CFamily/CFamilyEmbeddedSonarWayRulesTests.cs [hunk 6]
```diff
@@ -44,12 +44,14 @@ public class CFamilyEmbeddedSonarWayRulesTests
...
-        // Rule data for C-Family plugin v6.41 (build 60884)
-        private const int Active_C_Rules = 209;
-        private const int Inactive_C_Rules = 102;
+        // Rule data for C-Family plugin v6.47.0.62356
 
-        private const int Active_CPP_Rules = 435;
-        private const int Inactive_CPP_Rules = 165;
+        // Note: does not include MISRA rules - see #4604
+        private const int Active_C_Rules = 210;
+        private const int Inactive_C_Rules = 123;
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/CFamily/CFamilyEmbeddedSonarWayRulesTests.cs:51]: image

The 129 inactive rules for C overall are 123 inactive from the C family plugin + 6 inactive "Common" rules (see next comment).

**Comment by duncanp-sonar** [comment]
#### Standalone mode - MISRA rule explicitly enabled
- MISRA rule is explicitly enabled
- MISRA rule runs
- rule help is available in the IDE
- rule can be disabled
image


#### Connected mode with critical quality profile
- MISRA rules are run
- rule help is available in the IDE
- MISRA rules can't be disabled (since in ConnecedMode)

image

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**